mlflow/utils/environment.py

```diff
@@ -82,3 +89,61 @@ def _get_additional_pip_dep(conda_env):
 def _log_pip_requirements(conda_env, path, requirements_file="requirements.txt"):
     pip_deps = _get_additional_pip_dep(conda_env)
     _mlflow_additional_pip_env(pip_deps, path=os.path.join(path, requirements_file))
+
+
+def _get_pip_requirements(module_names):
+    module_to_packages = importlib_metadata.packages_distributions()
+    requirements = []
+    for module_name in module_names:
+        if module_name not in module_to_packages:
+            continue
+
+        for pkg in module_to_packages[module_name]:
+            req = "{}=={}".format(pkg, importlib_metadata.version(pkg))
+            if req in requirements:
+                continue
+
+            requirements.append(req)
+
+    return sorted(requirements)
+
+
+def _infer_pip_requirements(model_uri):
+    code_template = """
+import json
+import sys
+
+import mlflow
+
+def get_loaded_modules():
+   loaded_modules = set()
+   for mod_name in sys.modules.keys():
+       if not mod_name.startswith("_"):
+           loaded_modules.add(mod_name.split(".")[0])
+   return loaded_modules
+
+model = mlflow.pyfunc.load_model("{mode_uri}")
+
+print(json.dumps(list(get_loaded_modules())))
+"""
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        # Create a script to load the model
+        script_path = os.path.join(tmpdir, "load_model.py")
+        with open(script_path, "w") as f:
+            code = code_template.format(mode_uri=model_uri)
+            f.write(code)
+
+        # Run the script in a separate process and parse the output
+        stdout = subprocess.check_output(["python", script_path])
+        modules_names = json.loads(stdout.decode("utf-8"))
+        return _get_pip_requirements(modules_names)
+
+
+def _safe_infer_pip_requirements(model_uri, fallback):
```

Comment:
Nit: I think we should add the exception handling & fallback logic to `_infer_pip_requirements`, rather than defining a new method. Can we also make `_infer_pip_requirements` public? I.e. -> `infer_pip_requirements`? This will encourage users to leverage for additional use cases.

Reply:
Makes sense. I'll fix.